Allow users to specify generator class when using random behavior

joshuarwood opened the pull request:
I added an optional `rng` parameter to GDT core classes that provide random behavior, usually used for simulating event samples. The main benefits of this approach are:

1. It allows users to pass a Generator of their choice, including one initialized by a specific seed. Specifying the seed is necessary for reproducibility of unit tests and analysis results.
2. It enables support for multithreading as requested in #51. Users seeking multithreaded support can now follow the [numpy Generator guidance](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.spawn.html) and spawn child Generators for use in each thread:

```
import pymp
import numpy as np

from gdt.core.simulate.pha import PhaSimulator

rngs = np.random.default_rng(1).spawn(num_threads)

pha_sim = PhaSimulator(...)

with pymp.Parallel(num_threads) as p:
    for i in p.range(num_threads):
        pha_sim.rng = rngs[i]
        ...
```



Backwards compatibility is handled with a default value of None that results in the use of `np.random.default_rng()`. This provides similar random behavior seen in single threaded programs using the previous API.

Accepting this PR should close #51.

AdamGoldstein-USRA commented:
@joshuarwood It looks good, but can you add some documentation?  Probably just need a short note [here ](https://astro-gdt.readthedocs.io/en/latest/core/simulate/simulate.html)on initialization and multithreading.

joshuarwood commented:
@AdamGoldstein-USRA I think this is ready for review again. 

Changes since last time:
1. Some re-factoring to use a consistent `set_rng()` method across the API when updating the random generator after initialization
2. Updated unit tests with tests of the `set_rng()` method. These allow us to perform deterministic tests with a known seed in order to verify the exact output from simulation routines.
3. Updated doc pages for simulate/pha, simulate/tte, simulate/genorators with examples of seed setting and parallel processing

I ran the unit tests with python3.11 on my macbook and python3.12 on hippo. They seemed to complete successfully aside from the usual COBYLA fitter issue and some timeouts to HEASARC.
